REST API: Format should be passed via request header

Following REST principles the format is set by request header.

TEAM Engine should follow this common practice and define the requested format via request header instead of KVP.

Please also see: #217

If I send following requests (TEAM Engine runs on localhost:8081) with a web browser (Firefox) I just get an empty page and following stack trace is logged.
* http://localhost:8081/teamengine/rest/suites/wfs20/1.26/run?wfs=http://cite.deegree.org/deegree-webservices-3.4-RC3/services/wfs200?service=WFS%26request=GetCapabilities
* http://localhost:8081/teamengine/rest/suites/wms/1.23/run?capabilities-url=http://cite.deegree.org/deegree-webservices-3.4-RC3/services/wms130?service=WMS%26request=GetCapabilities
```
...
Error on line 1 column 1 of file:/usr/local/tomcat/temp/testng/d33845a5-a99c-4c8f-b81d-e03023b28a41/result.zip:
  SXXP0003: Error reported by XML parser: Content is not allowed in prolog.
Sep 08, 2017 2:57:10 PM com.sun.jersey.spi.container.ContainerResponse logException
SEVERE: Mapped exception to response: 500 (Internal Server Error)
javax.ws.rs.WebApplicationException: net.sf.saxon.trans.XPathException: org.xml.sax.SAXParseException; systemId: file:/usr/local/tomcat/temp/testng/d33845a5-a99c-4c8f-b81d-e03023b28a41/result.zip; lineNumber: 1; columnNumber: 1; Content is not allowed in prolog.
	at com.sun.jersey.core.impl.provider.entity.SourceProvider$SourceWriter.writeTo(SourceProvider.java:214)
	at com.sun.jersey.core.impl.provider.entity.SourceProvider$SourceWriter.writeTo(SourceProvider.java:172)
	at com.sun.jersey.spi.container.ContainerResponse.write(ContainerResponse.java:302)
...
Caused by: net.sf.saxon.trans.XPathException: org.xml.sax.SAXParseException; systemId: file:/usr/local/tomcat/temp/testng/d33845a5-a99c-4c8f-b81d-e03023b28a41/result.zip; lineNumber: 1; columnNumber: 1; Content is not allowed in prolog.
	at net.sf.saxon.event.Sender.sendSAXSource(Sender.java:379)
	at net.sf.saxon.event.Sender.send(Sender.java:164)
	at net.sf.saxon.IdentityTransformer.transform(IdentityTransformer.java:29)
	at com.sun.jersey.core.impl.provider.entity.SourceProvider$SourceWriter.writeTo(SourceProvider.java:208)
	... 29 more
Caused by: org.xml.sax.SAXParseException; systemId: file:/usr/local/tomcat/temp/testng/d33845a5-a99c-4c8f-b81d-e03023b28a41/result.zip; lineNumber: 1; columnNumber: 1; Content is not allowed in prolog.
	at org.apache.xerces.util.ErrorHandlerWrapper.createSAXParseException(Unknown Source)
	at org.apache.xerces.util.ErrorHandlerWrapper.fatalError(Unknown Source)
	at org.apache.xerces.impl.XMLErrorReporter.reportError(Unknown Source)
...
```

Expected is that EARL is returned.

--- a/teamengine-resources/src/main/resources/com/occamlab/te/earl_html_report.xsl
+++ b/teamengine-resources/src/main/resources/com/occamlab/te/earl_html_report.xsl
@@ -99,7 +99,7 @@
 						Result: <br />
 					    <div style="text-indent:50px;"> Number of conformance classes tested: <xsl:value-of select="count(//earl:TestRequirement)"/> </div>
 					    <xsl:variable name="no_of_cc" select="count(//earl:TestRequirement)" />
-					    <input type="hidden" id="noConformanceClass" name="noConformanceClass" value="{no_of_cc}" />
+					    <input type="hidden" id="noConformanceClass" name="noConformanceClass" value="{$no_of_cc}" />
 						<xsl:variable name="status">
 							<xsl:choose>
 								<xsl:when test="//earl:TestRequirement/cite:isBasic[text() = 'true'] and //earl:TestRequirement/cite:testsFailed[text() ='0']">

--- a/teamengine-spi-ctl/src/main/java/com/occamlab/te/spi/ctl/CtlExecutor.java
+++ b/teamengine-spi-ctl/src/main/java/com/occamlab/te/spi/ctl/CtlExecutor.java
@@ -109,7 +109,7 @@ public class CtlExecutor implements TestRunExecutor {
         Source results = null;
         try {
             File resultsFile = getResultsFile(getPreferredMediaType(testRunArgs), resultsDir.toString());
-            if(getPreferredMediaType(testRunArgs).endsWith("text/html")){
+            if(getPreferredMediaType(testRunArgs).endsWith("zip")){
               InputStream inStream = new FileInputStream(resultsFile);
               InputSource inSource = new InputSource(new InputStreamReader(inStream, StandardCharsets.UTF_8));
               results = new SAXSource(inSource);
@@ -170,9 +170,9 @@ public class CtlExecutor implements TestRunExecutor {
     	 // split out any media type parameters
         String contentType = mediaType.split(";")[0];
         String fileName = null;
-        if(contentType.endsWith("rdf+xml")){
+        if(contentType.endsWith("rdf+xml") || contentType.endsWith("rdf+earl")){
         	fileName = "earl-results.rdf";
-        } else if(contentType.endsWith("html")){
+        } else if(contentType.endsWith("zip")){
         	
         File htmlResult = HtmlReport.getHtmlResultZip(outputDirectory);
         fileName = "result.zip";
@@ -203,9 +203,11 @@ public class CtlExecutor implements TestRunExecutor {
             if (entry.getAttribute("key").equals("acceptMediaType")) {
             	if(entry.getTextContent().trim().equalsIgnoreCase("application/xml")){
                 mediaType = "application/xml";
-            	} else if(entry.getTextContent().trim().equalsIgnoreCase("text/html")){
-                    mediaType = "text/html";
-                	}
+            	} else if(entry.getTextContent().trim().equalsIgnoreCase("application/zip")){
+                    mediaType = "application/zip";
+                	} else if(entry.getTextContent().trim().equalsIgnoreCase("application/rdf+earl")){
+                        mediaType = "application/rdf+earl";
+                    	}
             }
         }
         return mediaType;

--- a/teamengine-spi-ctl/src/main/java/com/occamlab/te/spi/ctl/CtlExecutor.java
+++ b/teamengine-spi-ctl/src/main/java/com/occamlab/te/spi/ctl/CtlExecutor.java
@@ -109,7 +109,7 @@ public class CtlExecutor implements TestRunExecutor {
         Source results = null;
         try {
             File resultsFile = getResultsFile(getPreferredMediaType(testRunArgs), resultsDir.toString());
-            if(getPreferredMediaType(testRunArgs).endsWith("zip")){
+            if(getPreferredMediaType(testRunArgs).endsWith("text/html")){
               InputStream inStream = new FileInputStream(resultsFile);
               InputSource inSource = new InputSource(new InputStreamReader(inStream, StandardCharsets.UTF_8));
               results = new SAXSource(inSource);
@@ -172,7 +172,7 @@ public class CtlExecutor implements TestRunExecutor {
         String fileName = null;
         if(contentType.endsWith("rdf+xml")){
         	fileName = "earl-results.rdf";
-        } else if(contentType.endsWith("zip")){
+        } else if(contentType.endsWith("html")){
         	
         File htmlResult = HtmlReport.getHtmlResultZip(outputDirectory);
         fileName = "result.zip";
@@ -200,11 +200,11 @@ public class CtlExecutor implements TestRunExecutor {
         NodeList entries = testRunArgs.getElementsByTagName("entry");
         for (int i = 0; i < entries.getLength(); i++) {
             Element entry = (Element) entries.item(i);
-            if (entry.getAttribute("key").equals("format")) {
-            	if(entry.getTextContent().trim().equalsIgnoreCase("xml")){
+            if (entry.getAttribute("key").equals("acceptMediaType")) {
+            	if(entry.getTextContent().trim().equalsIgnoreCase("application/xml")){
                 mediaType = "application/xml";
-            	} else if(entry.getTextContent().trim().equalsIgnoreCase("html")){
-                    mediaType = "application/zip";
+            	} else if(entry.getTextContent().trim().equalsIgnoreCase("text/html")){
+                    mediaType = "text/html";
                 	}
             }
         }

--- a/teamengine-spi/src/main/java/com/occamlab/te/spi/executors/testng/TestNGExecutor.java
+++ b/teamengine-spi/src/main/java/com/occamlab/te/spi/executors/testng/TestNGExecutor.java
@@ -153,9 +153,9 @@ public class TestNGExecutor implements TestRunExecutor {
 		// split out any media type parameters
 		String contentType = mediaType.split(";")[0];
 		String fileName = null;
-		if (contentType.endsWith("rdf+xml")) {
+		if (contentType.endsWith("rdf+xml") || contentType.endsWith("rdf+earl")) {
 			fileName = "earl-results.rdf";
-		} else if (contentType.endsWith("html")) {
+		} else if (contentType.endsWith("zip")) {
 			File htmlResult = HtmlReport.getHtmlResultZip(outputDirectory);
 			fileName = "result.zip";
 		} else {
@@ -186,9 +186,11 @@ public class TestNGExecutor implements TestRunExecutor {
             if (entry.getAttribute("key").equals("acceptMediaType")) {
             	if(entry.getTextContent().trim().equalsIgnoreCase("application/xml")){
                 mediaType = "application/xml";
-            	} else if(entry.getTextContent().trim().equalsIgnoreCase("text/html")){
-                    mediaType = "text/html";
-                	}
+            	} else if(entry.getTextContent().trim().equalsIgnoreCase("application/zip")){
+                    mediaType = "application/zip";
+                	} else if(entry.getTextContent().trim().equalsIgnoreCase("application/rdf+earl")){
+                        mediaType = "application/rdf+earl";
+                    	}
             }
         }
         return mediaType;

--- a/teamengine-spi/src/main/java/com/occamlab/te/spi/executors/testng/TestNGExecutor.java
+++ b/teamengine-spi/src/main/java/com/occamlab/te/spi/executors/testng/TestNGExecutor.java
@@ -155,7 +155,7 @@ public class TestNGExecutor implements TestRunExecutor {
 		String fileName = null;
 		if (contentType.endsWith("rdf+xml")) {
 			fileName = "earl-results.rdf";
-		} else if (contentType.endsWith("zip")) {
+		} else if (contentType.endsWith("html")) {
 			File htmlResult = HtmlReport.getHtmlResultZip(outputDirectory);
 			fileName = "result.zip";
 		} else {
@@ -183,11 +183,11 @@ public class TestNGExecutor implements TestRunExecutor {
         NodeList entries = testRunArgs.getElementsByTagName("entry");
         for (int i = 0; i < entries.getLength(); i++) {
             Element entry = (Element) entries.item(i);
-            if (entry.getAttribute("key").equals("format")) {
-            	if(entry.getTextContent().trim().equalsIgnoreCase("xml")){
+            if (entry.getAttribute("key").equals("acceptMediaType")) {
+            	if(entry.getTextContent().trim().equalsIgnoreCase("application/xml")){
                 mediaType = "application/xml";
-            	} else if(entry.getTextContent().trim().equalsIgnoreCase("html")){
-                    mediaType = "application/zip";
+            	} else if(entry.getTextContent().trim().equalsIgnoreCase("text/html")){
+                    mediaType = "text/html";
                 	}
             }
         }

--- a/teamengine-spi/src/main/java/com/occamlab/te/spi/jaxrs/resources/TestResultLog.java
+++ b/teamengine-spi/src/main/java/com/occamlab/te/spi/jaxrs/resources/TestResultLog.java
@@ -34,7 +34,7 @@ import com.occamlab.te.spi.jaxrs.TestSuiteController;
 import com.occamlab.te.spi.jaxrs.TestSuiteRegistry;
 
 @Path("suites/{etsCode}/{etsVersion}/html/run")
-@Produces("application/zip")
+@Produces("application/zip, text/html")
 public class TestResultLog {
 
 	private static final Logger LOGR = Logger.getLogger(TestRunResource.class.getPackage().getName());

--- a/teamengine-spi/src/main/java/com/occamlab/te/spi/jaxrs/resources/TestResultLog.java
+++ b/teamengine-spi/src/main/java/com/occamlab/te/spi/jaxrs/resources/TestResultLog.java
@@ -34,7 +34,7 @@ import com.occamlab.te.spi.jaxrs.TestSuiteController;
 import com.occamlab.te.spi.jaxrs.TestSuiteRegistry;
 
 @Path("suites/{etsCode}/{etsVersion}/html/run")
-@Produces("application/zip, text/html")
+@Produces("application/zip; charset='utf-8'")
 public class TestResultLog {
 
 	private static final Logger LOGR = Logger.getLogger(TestRunResource.class.getPackage().getName());
@@ -45,7 +45,6 @@ public class TestResultLog {
     private HttpHeaders headers;
 	
 	@GET
-   // @Consumes({"application/html"})
     public Response handleHtmlGet(@PathParam("etsCode") String etsCode, @PathParam("etsVersion") String etsVersion ) throws IOException {
     	MultivaluedMap<String, String> params = this.reqUriInfo.getQueryParameters();
         Source results = executeTestRun(etsCode, etsVersion, params);

--- a/teamengine-spi/src/main/java/com/occamlab/te/spi/jaxrs/resources/TestRunResource.java
+++ b/teamengine-spi/src/main/java/com/occamlab/te/spi/jaxrs/resources/TestRunResource.java
@@ -43,7 +43,7 @@ import com.sun.jersey.multipart.FormDataParam;
  * @see <a href="http://jcp.org/en/jsr/detail?id=311">JSR 311</a>
  */
 @Path("suites/{etsCode}/{etsVersion}/run")
-@Produces({ "application/xml; charset='utf-8'", "application/rdf+xml; charset='utf-8'", "application/zip; charset='utf-8'" })
+@Produces({ "application/xml; charset='utf-8'", "application/rdf+xml; charset='utf-8'", "application/zip; charset='utf-8'", "application/rdf+earl; charset='utf-8'" })
 public class TestRunResource {
 
     private static final Logger LOGR = Logger.getLogger(TestRunResource.class.getPackage().getName());

